pepper-apis/src/main/java/org/broadinstitute/ddp/event/dsmtask/api/DsmTaskReceiver.java

```diff
@@ -0,0 +1,111 @@
+package org.broadinstitute.ddp.event.dsmtask.api;
+
+import static org.broadinstitute.ddp.event.dsmtask.api.DsmTaskData.ATTR_PARTICIPANT_GUID;
+import static org.broadinstitute.ddp.event.dsmtask.api.DsmTaskData.ATTR_STUDY_GUID;
+import static org.broadinstitute.ddp.event.dsmtask.api.DsmTaskData.ATTR_TASK_TYPE;
+import static org.broadinstitute.ddp.event.dsmtask.api.DsmTaskData.ATTR_USER_ID;
+import static org.broadinstitute.ddp.event.dsmtask.api.DsmTaskLogUtil.errorMsg;
+import static org.broadinstitute.ddp.event.dsmtask.api.DsmTaskLogUtil.infoMsg;
+import static org.slf4j.LoggerFactory.getLogger;
+
+
+import com.google.cloud.pubsub.v1.AckReplyConsumer;
+import com.google.cloud.pubsub.v1.MessageReceiver;
+import com.google.gson.Gson;
+import com.google.pubsub.v1.ProjectSubscriptionName;
+import com.google.pubsub.v1.PubsubMessage;
+import org.broadinstitute.ddp.util.GsonUtil;
+import org.slf4j.Logger;
+
+/**
+ * Receive and process DsmTask messages which are published by DSM to
+ * topic "dsm-to-dss-tasks" and fetched by subscriber registered by
+ * {@link DsmTaskPubSubConnectionCreator}.
+ *
+ * <p>When DsmTask is received it is processed by a {@link DsmTaskProcessor}
+ * which detected by a message attribute 'taskType' (from
+ * factory {@link DsmTaskProcessorFactory}.
+ *
+ * <p>After task is processed the result is sent back to DSM by
+ * publishing result message to topic "dss-to-dsm-results".
+ * The result message sending delegated to {@link DsmTaskResultSender}
+ */
+public class DsmTaskReceiver implements MessageReceiver {
+
+    private static final Logger LOG = getLogger(DsmTaskReceiver.class);
+
+    private final ProjectSubscriptionName projectSubscriptionName;
+    private final DsmTaskProcessorFactory dsmTaskProcessorFactory;
+    private final DsmTaskResultSender dsmTaskResultSender;
+
+    private final Gson gson = GsonUtil.standardGson();
+
+    public DsmTaskReceiver(ProjectSubscriptionName projectSubscriptionName,
+                           DsmTaskProcessorFactory dsmTaskProcessorFactory,
+                           DsmTaskResultSender dsmTaskResultSender) {
+        this.projectSubscriptionName = projectSubscriptionName;
+        this.dsmTaskProcessorFactory = dsmTaskProcessorFactory;
+        this.dsmTaskResultSender = dsmTaskResultSender;
+    }
+
+    @Override
+    public void receiveMessage(PubsubMessage message, AckReplyConsumer consumer) {
+        DsmTaskData dsmTaskData = parseMessage(message, consumer);
+        if (dsmTaskData != null) {
+            sendResponse(processDsmTask(dsmTaskData));
+            consumer.ack();
```

Comment:
Probably safer to `ack()` this between `processDsmTask()` and `sendResponse()`.  Otherwise, an error sending a complete message will trigger a stream of retries, which I don't think we want.

Reply:
Yup, I agree. If things look good and we finished processing it, we should `ack()`. And then it becomes our responsibility to send out the result message.

Reply:
Fixed